File: 221-maximal-square/maximal-square.ts
// the key note here is to notice the repeated work being done while checking for valid squares
// on on each matrix cell
// when noting this, u would go from a regualr brute-force  solution to an efficient DP solution
function maximalSquare(matrix: string[][]): number {
    
    const n:number = matrix.length;
    const m: number = matrix[0].length;
    const dp:number[][]=[];
    for (let i = 0; i < n; i++) {
        dp[i] = new Array(m).fill(0);
    }
    function countSquares(r:number,c:number):number {
        if(r>=n || c>=m){
            return 0;
        }
        if (dp[r][c] !== 0) {
            return dp[r][c];
        }
        let right = countSquares(r,c+1);
        let down = countSquares(r+1,c);
        let diagonal = countSquares(r+1,c+1);
        if(matrix[r][c]=="1"){
            // min because if any of the cells are 0 then we cannot make a square.
            dp[r][c]= 1 + Math.min(right,down,diagonal);
        }
        return dp[r][c];
    }
    countSquares(0,0);
    let maxSize = 0;

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            maxSize = Math.max(maxSize, countSquares(i, j));
        }
    }

    return maxSize * maxSize; 
};

File: 221-maximal-square/maximal-square.py
from functools import lru_cache
class Solution:
    # def maximalSquare(self, matrix: List[List[str]]) -> int:
 
        
        # memo = {}
        # R,C = len(matrix),len(matrix[0])
        # def dp(r,c):
        #     if r < 0 or r >= R or c<0 or c>=C:
        #         return 0

        #     if (r,c) in memo:
        #         return memo[(r,c)]
        #     down = dp(r+1,c)
        #     diagonal = dp(r+1,c+1)
        #     right = dp(r,c+1)
        #     best = 1 if matrix[r][c] == "1" else 0
        #     if best:
        #         best+= min(down,diagonal,right)
        #     memo[(r,c)] = best
        #     return memo[(r,c)]
        # dp(0,0)
        # return(max(memo.values())) ** 2
    def largestRectangleArea(self, heights: List[int]) -> int:
        n = len(heights)
        
        def next_lower():
            res = [n] * n
            s = []
            for i in range(n):
                while s and heights[s[-1]] > heights[i]:
                    res[s[-1]] = i
                    s.pop()
                s.append(i)
            return res
        
        def prev_lower():
            res = [-1] * n
            s = []
            for i in range(n-1, -1, -1):
                while s and heights[s[-1]] > heights[i]:
                    res[s[-1]] = i
                    s.pop()
                s.append(i)
            return res
        
        nl = next_lower()
        pl = prev_lower()
        
        best = 0
        for i in range(n):
            left = pl[i] + 1
            right = nl[i] - 1
            side = min(heights[i] ,(right - left + 1))
            best = max(best,side*side)
        
        return best
    
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        
        rows, cols = len(matrix), len(matrix[0])
        prefix = [[0] * cols for _ in range(rows)]
        
        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == '1':
                    prefix[i][j] = prefix[i][j-1] + 1 if j > 0 else 1
                else:
                    prefix[i][j] = 0
        
        max_area = 0
        
        for j in range(cols):
            heights = [prefix[i][j] for i in range(rows)]
            max_area_col = self.largestRectangleArea(heights)
            max_area = max(max_area, max_area_col)
        
        return max_area